### ISSUE TITLE
Backtest could be improved

### ISSUE THREAD

**Issue by fgnm** [Reiryoku-Technologies/Mida#104]
# Ticks and Periods
The actual backtest engine doesn't looks like to simulate real market data, what I'd like to have is something that can replicate as much as possible what happens in live, so have an alternation of ticks and periods, in the actual backtest instead when running `elapseTime` we have that all ticks in are fired before all periods https://github.com/Reiryoku-Technologies/Mida/blob/master/src/playground/MidaPlaygroundEngine.ts#L287-L328
This for me is wrong because in my strategy I want to do things when period is closed and other things when a new tick comes. The solution I propose is to use the property `MidaPeriod#ticks` to simulate better the market.
```ts
    async #processPeriods (periods: MidaPeriod[]): Promise<MidaTick[]> {
        const elapsedTicks: MidaTick[] = [];

        periods.sort((a: MidaPeriod, b: MidaPeriod): number => a.startDate.timestamp - b.startDate.timestamp);

        for (const period of periods) {
            let ticks: MidaTick[] | undefined = period.ticks;
            if (ticks == undefined || ticks.length == 0)
                ticks = [ tickFromPeriod(period, "close"), ];

            await this.#processTicks(ticks);
            await this.#onPeriodUpdate(period);

            /*if (period.isClosed) {
                await this.#onPeriodClose(period);
            }*/

            elapsedTicks.push(...ticks);
        }

        return elapsedTicks;
    }
```
In this way I can create from csv files a `MidaPeriod` array in which each period have its own ticks array and they are simulated almost exactly as live market (if a period doesn't have any tick fall back to the previous method).

# Trading System deadlock
Another issue I've found in the backtest engine is this call ` await this.#onPeriodClose(period);` (which I've commented). The `MidaTradingSystem` subscribe only to `tick` and `period-update` of the market watcher and not the `period-close`, but instead the close period callback of the system is fired by the system itself: https://github.com/Reiryoku-Technologies/Mida/blob/master/src/core/systems/MidaTradingSystem.ts#L448-L463

If the engine uses `waitFeedConfirmation` option it stuck on the first period because I've no way to proceed with the next feed. I've commented out just to get it working for me, I don't know which could be the best solution, please tell me :)

**Comment by Vasile-Peste** [comment]
Hello,
`This for me is wrong because in my strategy I want to do things when period is closed and other things when a new tick comes. The solution I propose is to use the property MidaPeriod#ticks to simulate better the market.`

You are absolutely right here. The fact is that the engine was built to be initially used either with ticks or periods, everyone agrees that it should work with both simultaneously.

Now, this is going to be a long post about how we can improve the backtesting engine so we are going discuss further improvements here. I will personally come back here with some proposed solutions taking into consideration also your solutions, thank you!

**Comment by fgnm** [comment]
Thank you for all the support, I'll keep update this issue with more improvements for the backtest engine, reporting the one already discussed on discord to keep track of it :)

# Performances
There are some performances issues that heavily impact on backtest engine performances. With huge datasets it's not practical to perform a backtest in a reasonable time. Two major bottlenecks I've found are the `MidaPlaygroundEngine` and in `MidaDecimal`.

For the first one, the abuse of `filter` method for the array slows down a lot. Especially for the `getOpenPositions` function. The proposed fix here is to replace with simple for-loop. i.e.:
```ts
    public async getOpenPositions (): Promise<MidaPosition[]> {
        let openPositions: MidaPosition[] = [];

        for (let [id, position] of this.#positions) {
            if (!position.volume.isZero()) openPositions.push(position);
        }

        return openPositions;
    }
```

The `MidaDecimal` has a huge impact on backtest performances, in particular the `decimal` function, widely used in the backtest engine, always allocate new objects (increasing GC time) and the `getParts` function requires a lot of time to compute. A partial fix is to add a `isZero` function to help reduce the creation of uselss objects in some scenarios (like `getOpenPositions`):
```ts
    public isZero (): boolean {
        return this.#value === 0n;
    }
```
The the major issue I wasn't able to fix.

# Report generation and equity
In a backtest we usually wants to know which positions where opened, when closed, drawdown, etc. Would be very nice to have a final report for all positions happened during the backtest and some stats.

Also I noticed that the `TradingAccount#getEquity` isn't very useful, because it just tries to sell the assets but doesn't look at real position opened, its behavior in backtest engine is completely different then cTrader for example.

**Comment by Vasile-Peste** [comment]
So to have a brief summary, we have to:
    1. Create a way to elapse a backtest using a file stream as data source;
    2. Avoid using .filter/forEach and any functional
    3. See what we can do to improve the MidaDecimal processing times (starting from the isZero proposal)
    4. Process and emit ticks and periods in parallel

A plus but not required for this stage is generating a report of the backtest, can be calculated manually for now
I have created a PR where we will do our best to address the points above https://github.com/Reiryoku-Technologies/Mida/pull/113
